src/management-system-v2/lib/user-preferences.ts

```diff
@@ -54,6 +42,119 @@ const defaultPreferences = {
 }; /* as const */ /* Does not work for strings */
 
+const partialUpdate = (
+  defaultPreferences: PreferencesType,
+  currentPreferences: PreferencesType,
+): PreferencesType => {
+  const updateArray = (defaultArray: any[], currentArray: any[]): any[] => {
+    /* In case default array is empty */
+    if (defaultArray.length === 0) return currentArray;
+
+    /* Shape-Switch */
+    let shapeIsDifferent = false;
+
+    /* just take first element of array, since only the inner shape and not the values are interesting */
+    const defaultValueSample =
+      defaultArray[0]; /* NOTE: Assumes that every default array has at least one entry! (Check above) */
+
+    const updatedArray = currentArray.map((currentValue) => {
+      /* For speed up */
+      if (shapeIsDifferent) return currentValue;
+
+      /* Default elment is an array */
+      if (Array.isArray(defaultValueSample)) {
+        /* Currently saved is not an array */
+        if (!Array.isArray(currentValue)) {
+          shapeIsDifferent = true;
+          return defaultValueSample;
+          /* Currently saved is an array */
+        } else {
+          return updateArray(defaultValueSample, currentValue);
+        }
+        /* Default elment is an an object */
+      } else if (typeof defaultValueSample === 'object' && defaultValueSample !== null) {
+        /* Currently saved is not an object */
+        if (
+          typeof currentValue !== 'object' ||
+          currentValue === null ||
+          Array.isArray(currentValue)
+        ) {
+          shapeIsDifferent = true;
+          return defaultValueSample;
+          /* Currently saved is an object */
+        } else {
+          /* Shape is conform with default shape (first element) */
+          try {
+            const result = updateObject(defaultValueSample, currentValue);
+            return result;
+            /* Shape is not conform with default shape */
+          } catch (error) {
+            shapeIsDifferent = true;
+            return defaultValueSample;
+          }
+        }
+        /* Everything else */
+      } else {
+        return currentValue;
+      }
+    });
+
+    return shapeIsDifferent ? defaultArray : updatedArray;
+  };
+
+  const updateObject = (
+    defaultObject: Record<string, any>,
+    currentObject: Record<string, any>,
+    root: boolean = false,
+  ): Record<string, any> => {
+    let updatedObject: Record<string, any> = {};
+    /* Remove entries not present in default */
+    for (const key in defaultObject) {
+      updatedObject[key] = currentObject.hasOwnProperty(key)
+        ? currentObject[key]
+        : defaultObject[key];
+    }
+
+    for (const key in defaultObject) {
+      /* Not present yet */
+      if (!currentObject.hasOwnProperty(key)) {
+        /* For nested objects -> replace them with default as something has changed */
+        if (!root) throw new Error(`Mismatch in nested object in Preferences: ${key} is missing!`);
```

Comment:
The comment and code don't seem to match. Is it supposed to throw there?

Reply:
I used throwing an error here to backtrack, so in other words:
During updating an object it might become clear, that the object itself needs to be replaced with the default, as it is not updateable, as the loop just updates/ replaces elements in the object and not the complete object, this was a way of telling the calling block to use the default. 

Changed it to just breaking the loop and then returning the default - I think it is clear now.

Reply:
Yeah, I just meant that there was only a try catch in the updateArray and not the updateObject.